### ISSUE TITLE
What is a sort injection?

### ISSUE THREAD

**Issue by ttuegel** [kframework/kore#681]
What is a sort injection?

Given these syntax definitions,
```
syntax BC ::= "B" | "C"
syntax A ::= "A" | BC
```
the sort `A` in Kore contains the constructor `A`, and we have subsort axioms for injections of  `inj{BC, A}(B)` and `inj{BC, A}(C)`. 

**Q1(a)**: Are these injections in the image of some constructor, or are they themselves constructor-like?

**Q1(b)**: If the injections are in the image of some constructor, which constructor?

The Haskell backend currently treats them as constructor-like; for example,
```
\ceil(A ∧ inj{BC, A}(B)) = ⊥
```

On the other hand, the no-junk constructor axioms for sort `A` do not mention them; that is, we have
```
axiom A
```
instead of
```
axiom A ∨ (∃ x. inj{BC, A}(x))
```

The answer to **Q1** has implications for how we treat overloaded symbols. Consider these definitions, adapted from #680:
```
syntax Number ::= Int | Float

syntax IType ::= "i32" | "i64"
syntax FType ::= "f32" | "f64"
syntax  Type ::= IType | FType

syntax IVal ::= overloaded(IType, Int) [klabel(overloaded)]
syntax FVal ::= overloaded(FType, Float) [klabel(overloaded)]

syntax Val ::= IVal | FVal
             | overloaded(Type, Number) [klabel(overloaded)]
```
Each `overloaded` symbol is regarded as a constructor in its own sort. Because all three `overloaded` symbols have the same `klabel`, the frontend generates overloading axioms such as
```
axiom overloaded_Val(inj{IType, Type}(T), inj{Int, Number}(X)) =
        inj{IVal, Val}(overloaded_IVal(T, X))
```
If `inj{IVal, Val}` is constructor-like, this axiom violates the no-confusion axiom for constructors in `Val`! The intent of such overloading is that the `overloaded` symbol in `Val` is merely an abstraction for the injection of the `IVal` and `FVal` symbols, but all three symbols are given constructor axioms. We could regard `overloaded_Val` as a (partial) function instead, but the other backends cannot handle functions on the left-hand side of rules.

**Q2**: What is `overloaded_Val`?

**Comment by xc93** [comment]
These injections are constructor-like. The set of all constructors of sort `A` contains the terminal `A` and the injection `inj{BC, A}`. This means that when stating the (No Junk) axioms, we need to take the injection into account. In addition, these injections are also injective functions. 

In terms of (No Confusion), I can see why it is a more tricky issue. Because injections need to satisfy the triangle property: `inj{A,C}(x) = inj{B,C}(inj{A,B}(x))`, their (No Confusion) axioms are unevitably more complex than the normal constructors. (Note that the triangle property is already in conflict with the (No Confusion) axiom, if we regard `inj{A,C}` and `inj{B,C}` as two normal constructors of sort `C`.)

Note that we have a similar issue in domains that contain constructors satisfying some additional "theory axioms". One typical example is sort `Map` for finite maps, which has constructors `emp` for the empty map, `_|->_` for singleton maps, and `_*_` to merge two maps into a bigger map. And `_*_` satisfies the variant of (No Confusion) that takes its associativity and commutativity, and the axiom `x|->y * x|->z = emp` into account, which is unevitably more complex than the normal constructor case.

Back to your question: It is in general tricky to generate (No Confusion) for injections. But it is easy to generate (No Confusion) for normal constructors, and, in the case of (No Confusion Different Constructor), for any pair of a normal constructor and an injection. I believe these axioms will be enough to drive the unification process, which, I guess, would be the main use of (No Confusion) in backends. For (No Confusion Different Constructor) for a pair of two injections, this boils down to properly dealing with subsorting in general and, in particular, the triangle property. 

To sum up, my suggestion for this issue would be to forget about generating a complete set of (No Confusion) axioms, but only those that are needed, (for unification, e.g.). It is easy to define axioms that specify positive statements (e.g., when two things are equal) and hard to specify negative statements (e.g., when two things are unequal as in (No Confusion)).

**Comment by ttuegel** [comment]
> But it is easy to generate (No Confusion) for normal constructors, and, in the case of (No Confusion Different Constructor), for any pair of a normal constructor and an injection.

This is only easy if there is no overloading! If one of the constructors in the higher sort overloads a constructor in a lower sort, what should be the "no confusion" axiom? Or even worse: what should be the "no confusion" axiom if a constructor in a higher sort overloads one of many constructors in a lower sort?

**Comment by xc93** [comment]
Yes. You are right on that. It's only easy, I guess, if there's no overloading; and if there is, then it is only easy if both constructors are normal constructors.

**Comment by grosu** [comment]
Thanks for starting this discussion, @ttuegel .  Yes, it is tricky.  In K, we are talking about constructors modulo axioms like assoc, comm, and inj-triangles.  Constructors modulo assoc and comm are pretty understood from previous languages like OBJ and Maude, but other axioms, like injection triangles, are less well understood.  And later on we should also consider context and their axioms.  I think we need to write a research paper about it, in fact.

I agree with all the points that @xc93 made above.  Also, more importantly, that we should probably not get stuck with having a perfect axiomatization that allows us to proof-check everything we do algorithmically with the current version of ML that K implements.  I know I should be the last person on the planet to say that, but here is my rationale: I am convinced that we will switch to AppML eventually, where things are *so much* cleaner and simpler.  Then all this business with injections and parametric symbols will dissolve and either we'll not have at all these problems or we'd have to translate them into something else.

Let me re-iterate it publicly here that I am pretty opposed to order-sorted algebra (OSA) as a "foundation of things" these days.  I started my PhD on OSA with Goguen, the inventor of OSA (Meseguer came into the game a bit later).  I showed that OSA can be organized as an institution, which was pretty technical.  Then I also proved the orthogonality-implies-local-confluence theorem for OSA at Goguen's request, which was one of the most complex things I've ever proved.  Then I used OSA in several of my papers and for 20 years with Maude.  My conclusion after all this years is that the complexity that OSA adds is *not worth* to build foundations on it.  OSA is great for algorithmic advances, like we are also doing here, but in terms of foundations, to explain things and develop theories, AppML is in my view the right answer after all these years of search.  All the above can be pretty easily explained with AppML.  @xc93 , can you please do it?

So my suggestion is the following: feel free to do whatever you need to algorithmically, provided that you can explain it in terms of AppML.  For now, explaining it informally and weaving hands if fine; later on  (in 1 year or so), we will generate AppML proof objects and then we will need to make everything rigorous.

**Comment by ttuegel** [comment]
> we should probably not get stuck with having a perfect axiomatization

I agree, but I can't even say using _informal_ language what these things (sort injections and overloads) should mean. Sort injections without overloading are clear now, thanks to @xc93's comments. And while we don't need a complete axiomatization, we should at least be sure that the frontend does not emit *wrong* axioms. It's fine if we have to use built-in reasoning where the axioms are not sufficient, but we do use some axioms internally (and also send some to Z3).

**Comment by xc93** [comment]
As said in @grosu's comments, a lot of these issues with subsort overloading and proper axiomatization will go away in AppML. Let me give an example here. Firstly, the biggest game changer in AppML is that we drop all sorts and, instead, we deal with sorts by using constant symbols that represent their names (such as `Int, Float, Number, Type,...`) and a special symbol `[[_]]`, called _inhabitant symbol_, which, when applied to a sort name (e.g., `Int`), the pattern (e.g., `[[Int]]`) represents the inhabitant set of that sort. Subsorting relation is then easily captured by the following axioms:
```
[[Number]] = [[Int]] \/ [[Float]]
[[Type]] = [[IType]] \/ [[FType]]
[[Val]] = [[IVal]] \/ [[FVal]]
```
The function `overloaded`, which is (subsort) overloaded, can be easily captured in AppML by defining just one function, say `overloaded`, with the following (Arity) axioms that specify its behavior over different sort universes (Note that I write function application `f(x,y)` as `f x y` because that's more preferable in AppML; it's just a matter of notation):
```
overloaded [[Int]] [[IType]] ⊆ [[IVal]]
overloaded [[Float]] [[FType]] ⊆ [[FVal]]
overloaded [[Number]] [[Type]] ⊆ [[Val]]
```
No injections are needed at any place. It's just sort names, their inhabitant, and one function `overloaded`, with the natural axioms specifying the subsort relation and the arity function. Other axioms such as (No Junk) and (No Confusion) are also straignforward to define. As an example:
```
(No Junk)  [[IVal]] = overloaded [[Int]] [[IType]]   // that is, we strengthen "⊆" to "=" in the (Arity) axiom
(No Confusion) overloaded x t = overloaded x' t' -> x=x' /\ t=t'
(No Confusion) [[IVal]] ∩ [[FVal]] = ∅
```

**Comment by ttuegel** [comment]
@xc93 We (myself with @virgil-serbanuta) have another question, regarding sort injections where there is a diamond relation between sorts, i.e.
```
Sub ⊂ A
Sub ⊂ B
A ⊂ Super
B ⊂ Super
```
The term `inj{Sub, Super}(φ)` is abiguous: it could refer to `inj{A, Super}(inj{Sub, A}(φ))` or `inj{B, Super}(inj{Sub, B}(φ))`. What is the resolution of this ambiguity? The two patterns cannot be the same if there is a no-confusion-like axiom for injections.

**Comment by ttuegel** [comment]
Based on [the response](https://runtimeverification.slack.com/archives/CC360GUTG/p1563609925003000), I think there is nothing more to do here right now.
